File: javascript/205.isomorphic-strings.js
/*
 * @lc app=leetcode id=205 lang=javascript
 *
 * [205] Isomorphic Strings
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function (s, t) {
  if (s === t) return true;

  // Solution 1:
  // const s_chars = {};
  // const t_chars = {};
  // for (let i = 0; i < s.length; i++) {
  //   if (!s_chars.hasOwnProperty(s[i])) {
  //     s_chars[s[i]] = t[i];
  //   } else if (s_chars[s[i]] !== t[i]) {
  //     return false;
  //   }
  //   if (!t_chars.hasOwnProperty(t[i])) {
  //     t_chars[t[i]] = s[i];
  //   } else if (t_chars[t[i]] !== s[i]) {
  //     return false;
  //   }
  // }

  // Solution 2:
  for (let i = 0; i < s.length; i++) {
    if (s.indexOf(s[i], i + 1) !== t.indexOf(t[i], i + 1)) return false;
  }

  return true;
};
// @lc code=end

File: rust/205.isomorphic-strings.rs
/*
 * @lc app=leetcode id=205 lang=rust
 *
 * [205] Isomorphic Strings
 */

// @lc code=start
impl Solution {
    pub fn is_isomorphic(s: String, t: String) -> bool {
        if s == t {
            return true;
        }

        // Solution 1: using HashMap
        use std::collections::HashMap;
        let mut s_map: HashMap<char, char> = HashMap::new();
        let mut t_map: HashMap<char, char> = HashMap::new();
        for (a, b) in s.chars().zip(t.chars()) {
            match (s_map.get(&a), t_map.get(&b)) {
                (Some(&c), Some(&d)) => {
                    if c != b || d != a {
                        return false;
                    }
                }
                (None, None) => {
                    s_map.insert(a, b);
                    t_map.insert(b, a);
                }
                _ => return false,
            }
        }

        // Solution 2: using Vec
        // let mut pairs: Vec<(char, char)> = Vec::new();
        // for (a, b) in s.chars().zip(t.chars()) {
        //     if let Some((c, d)) = pairs.iter().find(|(u, v)| (*u == a || *v == b)) {
        //         if a != *c || b != *d {
        //             return false;
        //         }
        //     } else {
        //         pairs.push((a, b));
        //     }
        // }

        return true;
    }
}
// @lc code=end
